src/apis/ActionSheetIOS.re

```diff
@@ -1,36 +1,48 @@
-type options;
-[@bs.obj]
-external options:
-  (
-    ~options: array(string),
-    ~cancelButtonIndex: int=?,
-    ~destructiveButtonIndex: int=?,
-    ~title: string=?,
-    ~message: string=?,
-    ~tintColor: Color.t=?,
-    unit
-  ) =>
-  options =
-  "";
+type options = {
+  options: array(string),
+  cancelButtonIndex: option(int),
+  destructiveButtonIndex: option(int),
+  title: option(string),
+  message: option(string),
+  tintColor: option(Color.t),
+};
```

Comment:
Only values returned from an API (e.g. as callback params) should be converted to records. Values passed to an API (options) should not be converted to records, because records cannot have optional fields.

With the current bindings, you can do

```reason
let options = ActionSheetIOS.options(~options=[|"A", "B", "C"|], ());
```

(all the other props being optional). If converting to records you **won't** be able to do

```reason
let options = { ActionSheetIOS.options: [|"A", "B", "C"|] };
```

but would have to supply all the remaining props with value `None` which is very inconvenient.

Reply:
In other words, we should be using records where we are currently using `Js.t` objects, but not where we are currently using abstract types.

Reply:
> So you would need to check each case individually to be sure it is safe and the API never returns null before converting it to option.

I am happy to do this. How would you go about doing that?

Reply:
@cknitt: You can create the same constructor function (that currently  creates objects) to  create records with default value None.

Reply:
@jfrolich Can you show an example to be sure about what you mean?

Reply:
Basically have a function with optional labels that construct the record.

Reply:
@jfrolich @MoOx 

There is a subtle difference between optional params in a `bs.obj` creation function and `option` fields in a record. With `bs.obj`, the optional fields will be omitted from the JS object if not specified.

However,

```re
let options = {
  options: [| "A", "B" |],
  cancelButtonIndex: None,
  destructiveButtonIndex: None,
  title: None,
  message: None,
  tintColor: None,
};
```
will give you the JS object

```js
{
  options: [ 'A', 'B' ],
  cancelButtonIndex: undefined,
  destructiveButtonIndex: undefined,
  title: undefined,
  message: undefined,
  tintColor: undefined,
};
```

So

1. This difference in behavior can lead to subtle bugs in some situations,
2. what you construct with the `bs.obj` creation function strictly speaking has *not* the same type as the record, as a record always needs to have all the fields, and
3. this may lead to code size issues.

I am therefore very hesitant to provide these record types for "input" params.

Reply:
@cknitt: Yes there might be an issue with javascript code that will check for existence of a key instead of the null value. However I think that will be an edge case (need to test though!)

In the case when there is a check if keys exists on an object it's also possible to convert the record to Js.Json.t and only keep the non-null record fields (or just keep the `bs.obj` function). But when the interop works with records it provides quite an ergonomic interface.

Reply:
Hi folks. Sorry for the prolonged absence. I am back now. If I am reading this correctly, we were considering converting our `input` types back to `bs.obj`'s. Do we still want to do this?

Reply:
Great you are back @idkjs! I think it could be records unless input API's rely on existence of keys.

Reply:
I can confirm that using the record syntax gets requires the following `option` code:

```reason
let options: ReactNative.ActionSheetIOS.options = {
  options: buttons,
  cancelButtonIndex: Some(4),
  destructiveButtonIndex: Some(3),
  tintColor: Some("green"),
  title:None,
  message:None
};
```


Which compiles to 
```js
var options = {
  options: buttons,
  cancelButtonIndex: options_cancelButtonIndex,
  destructiveButtonIndex: options_destructiveButtonIndex,
  title: undefined,
  message: undefined,
  tintColor: options_tintColor
};
```

and runs fine. Don't know if this answers the question. I think going with records is the way to go for what it is worth. At the very least, you will always be aware of what the `js` code is doing. Examples running against `bs.obj` [idkjs/ActionSheetIOSDemo](https://github.com/idkjs/ActionSheetIOSDemo). Against records in a minute.

Reply:
What do you folks want to do with this? `bs.obj` or `records`?

Reply:
I think we definitely don't want to type

```reason
let options: ReactNative.ActionSheetIOS.options = {
  options: buttons,
  cancelButtonIndex: None,
  destructiveButtonIndex: None,
  tintColor: None,
  title:None,
  message:None
};
```

When you can just ask people to write this

```reason
let options = ActionSheetIOS.options(~options=buttons, ());
```

So we must keep `bs.obj` for now.